plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2055,15 +2076,24 @@ void producer_plugin_impl::produce_block() {
       return sigs;
    } );
 
-   chain.commit_block();
+   if ( blockvault_plug->get() != nullptr ) {
+      std::promise<bool> p;
+      std::future<bool> f = p.get_future();
+      blockvault_plug->get()->async_propose_constructed_block({pending_blk_state->block->block_num(), pending_blk_state->block->timestamp},
+                                                               pending_blk_state->dpos_irreversible_blocknum,
+                                                               pending_blk_state->block, [&p](bool b) {
+         p.set_value( b );
+      });
+      if (!f.get())
+         throw rejected_by_blockvault{};
```

Comment:
I think it is worth using `EOS_ASSERT` here instead of a raw throw with a struct. Declare a new exception type and use it.

Reply:
The reason I chose not to use EOS_ASSERT and fc derived exception was because I didn't want to log the exception detail.  Logging the exception detail just makes it looks like error as you have mentioned. If we don't log the exception detail, I don't see why not just use a simple struct based exception. This exception should never be propagated outside this file, declare a new exception in `chain/exceptions.hpp` just make you recompile more than 300 files.

Reply:
`block_validate_exception` would be appropriate. It is an exceptional condition that is an error. My point is that the error is with the block not with the blockvault as was reported before.